### DISCUSSION TITLE
Graphics interfaces for sicmutils

### DISCUSSION THREAD

**Discussion by alexgian** [sicmutils/sicmutils#301] [Ideas]
This is an interesting and challenging aspect of sicmutils.  In SICM, GJS uses a number of illustrative diagrams, all of which can be created fairly easily using elementary plotting routines like points and lines.  However the wealth of utilities available nowadays, both to JVM and to Javascript environments suggests that a lot more could be done to enhance the educational value of the system.

MIT-Scheme itself relies on a very simple library 《see Source Links, below》, which, besides the basic primitives, allows interfaces provided by the X system to be selected, queried for abilities and accessed.  Although, happily, we are not tied to the X system any more, we could still gainfully keep some of these ideas:  We may be deploying sicmutils on the JVM, or on the browser, via Clojurescript.  It would be good if the system could query what abilities, what "canvas" (in the most general sense) is available to it.

Not only that, but even approaches to graphical representation can vary a lot.  For instance, most of the animations in the book are created naively, using a loop to render them frame by frame.  More modern systems will have their own Draw() and Update() functions, handled by the system itself.  I think it is important that whatever systems are used for rendering will support both approaches, the former for backward compatibility with the text, the latter to allow more creative ideas to proceed unhampered by older idioms.  This would be particularly important too, when tactile interfaces, such as those found in "tablet" devices are targeted.

On top of the MIT-Scheme graphics primitives, GJS's scmutils uses a couple of simple libraries that provide the basic primitives you'd expect 《see Source Links, below》.
I think it would be a good idea that any graphics interface we make available, can, at least run the examples from the book - verbatim.  That, or as a bare minimum, when queried, declare that it cannot and would the user kindly use a different interface.

**Comment by alexgian** [comment]
Continuing with the above idea, we could, for instance have a "bare-minimum" plotter that only covers the requirements of the book(s) and the basic GJS library.  They could both (Clojure and Clojurescript version) have the same name and implement the same interface, even though they will probably be implemented by totally different code.  However, this bare-minimum functionality could also be available through a more advanced plotter (canvas or device or library or whatever you want to call it) running either on JVM or Javascript.  However, the more sophisticated system might render it better or prettier or faster or with more control options.  In this case the user would just specify a change of library from "Barebones" to "Muchbetter", but still be able to type the code straight out if the book.

Now, once the user is comfortable with a bit more freestyling, either exploring a subject more fully to their own satisfaction, or perhaps creating a tutorial or demo to share, then they can use the more advanced features available, and use SICMutils as the library.  It would be a matter of the creativity of the user.

**Reply by sritchie** [reply]
I'm very pro this discussion, @alexgian . I did some experimenting with Vega-Lite at this notebook: https://github.com/scicloj/notespace-sicmutils-example/blob/main/src/notespace_sicmutils_example/ex1_44.clj#L210

That produced the following output: https://scicloj.github.io/notespace-sicmutils-example/docs/notespace-sicmutils-example/ex1-44/

You can see interactive charts a good way down the page:

![image](https://user-images.githubusercontent.com/69635/109548379-d3f3a380-7a89-11eb-9bc9-4ec67449f83b.png)

I don't actually like the Hanami library used for this, but it does indeed seem like Vega-Lite ([see this video](https://www.youtube.com/watch?v=9uaHRWj04D4) is a modern, lovely choice for some subset of the things that GJS and JW do in the book.

Do you have much experience here? What do you think?

I had also done some Quil work a while back for animations, as I think you'd seen: https://github.com/sicmutils/maturin

And of course the system really wants to have the ability to plot interactive graphs of arbitrary functions, in service of its "mini-Mathematica" aspirations.

Would you like to lead this design + implementation effort, @alexgian ?

**Reply by alexgian** [reply]
I've just looked a bit at the stuff you did with the Hanami interface on scicloj/notespace/Ex1.44 and even though it looks a bit hairy initially, I don't think it's too bad.  I'll have to mess about a bit with that declarative interface / JSON approach, and no, I've never done any graphics like that before, so no real experience, but I think it will be OK.  That Vega-lite system seems to offer so much ready out-of-the-box that I'm wavering again.  It shouldn't be too difficult to map Sussman's routines/approach to it, even with just an ad hoc stab at it,initially.  I mean, you've done it already...

You say you don't like Hanami.  Any particular reason?  What are the other options?  Oz?  I haven't looked at that yet.  I am assuming the Vega code will be the same in both cases, just plumbed in differently - I may be wrong though.

I'll try give it a run soon and see how well it works for interactive animation.  (What we said about tablet-style interfaces).  If it covers this well enough, then Quil might not even be necessary.

**Reply by sritchie** [reply]
@alexgian all I meant was that SICMUtils probably doesn't need a dependency like Hanami to generate the templates that it will offer. If someone wants to use functions from

- Hanami
- https://github.com/applied-science/waqi
- https://github.com/applied-science/hubble
- https://github.com/metasoarous/oz

that should all be fine and lovely!

@alexgian , I've also cleaned up my Quil animation demos so they work with modern SICMUtils and have notes in the `core.clj` namespace about what I was trying to do: https://github.com/sicmutils/maturin

Blake Miller (in Clojurians Zulip) is also looking at this stuff. Let me ping him and get him talking here on the thread.

Thanks @alexgian !

**Comment by alexgian** [comment]
There are a number of systems available for this use.  The following is an incomplete list:

- Simple AWT in Java, or something just as simple in JS with HTML5.

- Processing or one of its variants, P5js, etc, or Quil if you want to design in Clojure!  With this, it seems to me that the advantage is that the codebase for both JVM and JS would be very similar, and the system already has some popularity and traction.  Also, in a JVM environment it could work without need for a web-server and browser. I have already started to explore it a bit.  On the minus side, there does not seem to be much in ready-made graphing toooling available, although Sam has made a start with maturin.

- D3, vega and vega-lite.  Very strong JS system for mathematical representation, and nicely layered, so at the top vega lite provides a simple graphing interface.  The more you want to do, the deeper you may have to go, right down to D3.  I think JVM would need to access it via a web setup (websocket?) though. Sam has already demonstrated Ex. 1.44 on it

- The Geogebra system, with geometry, CAS and spreadsheet functionality and extreme ease in designing linkages quickly could be used for demos.  It would have to be accessed by some sort of RESTful interface though.

- Mini Unity-style game creators and modelers like three.js, babylon, etc.  A lot of potential here, but probably something to look at once the basics are in place.

I also have no idea how easily such subsystems may be incorporated in notebooks like Jupyter et al. I imagine there won't be any major problem but if anyone knows of any such I'd like to hear about them.

**Reply by sritchie** [reply]
Whoops, of course you have everything here that I hinted at above... sorry about that!! Ignore my repetition.

I think there is no problem integrating this stuff with modern notebooks. I added a list of environments we may want to test against here: https://github.com/sicmutils/sicmutils/blob/master/doc/basics/using_sicmutils.adoc

**Comment by alexgian** [comment]
«Source Links» (from above)

I cannot find any MIT-Scheme code unzipped on the web, but the file with the X devices and primitives is src/x11/x-11device.scm. It is not really that important, but I include part of it it in the gist below to show what MIT-Scheme provides as a basis.
https://gist.github.com/alexgian/15a0c3f5c0b1f34bc515dacbf13c6a7f

Upon this GJS builds his plotting routine, which is at the core of the graphics he creates, in his usual inimitable style.  Again  there is no reason to do it like this, especially in settings that may supply primitives quite different from those of X11, but at this stage it is probably useful at least to note the interface that is emerging. See:
https://github.com/Tipoca/scmutils/blob/master/src/gjsgraph/plot.scm

Finally GJS builds the following grapher, based on the previous plotter, and, to the extent that this is used in the book(s) I think it should be supported as much as possible.
https://github.com/Tipoca/scmutils/blob/master/src/gjsgraph/graph.scm

**Reply by sritchie** [reply]
I don't know that he uses it much in code the user types, which is great. He definitely uses it to generate the charts; this is wonderful, because we can do it in our own way without breaking the contract with the user.

@alexgian do you want to lay out the menu of things we want to support? Our roadmap.

- dynamical systems plots, generated by `evolve`
- arbitrary function plotting, as with `graph.scm`
- animating systems linked to these plots

What else?
